network/p2p/dnsaddr/resolve_test.go

```diff
@@ -109,3 +110,42 @@ func TestMultiaddrsFromResolverDnsFailure(t *testing.T) {
 	assert.Empty(t, maddrs)
 	assert.ErrorContains(t, err, "always errors")
 }
+
+type mockController struct {
+}
+
+func (c mockController) Resolver() Resolver {
+	return selfResolver{}
+}
+
+func (c mockController) NextResolver() Resolver {
+	return nil
+}
+
+type selfResolver struct {
+}
+
+func (r selfResolver) Resolve(ctx context.Context, maddr multiaddr.Multiaddr) ([]multiaddr.Multiaddr, error) {
+	return []multiaddr.Multiaddr{maddr}, nil
+}
+
+// TestIterate ensures the Iterate() does not hang in infinite loop
+// when resolver returns the same dnsaddr
+func TestIterate(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	dnsAddr := "/dnsaddr/foobar.com"
+	require.True(t, isDnsaddr(multiaddr.StringCast(dnsAddr)))
+	ma, err := multiaddr.NewMultiaddr(dnsAddr)
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		Iterate(
+			ma,
+			mockController{},
+			func(dnsaddr multiaddr.Multiaddr, entries []multiaddr.Multiaddr) error { return nil },
+		)
+		return true
```

Comment:
Ideally this would check the error returned by `Iterate`

Reply:
why? I want to check no infinite loops and not resolution errors

Reply:
You made the decision to return an error when max hops are reached. In my view, it only makes sense to test that return value here, rather than ignoring it